cql3/expr/prepare_expr.cc

```diff
@@ -1046,6 +1047,183 @@ ::shared_ptr<assignment_testable> as_assignment_testable(expression e) {
+            on_internal_error(expr_logger, format("get_lhs_receiver: unexpected expression: {}", other));
+        },
+    }, prepared_lhs);
+}
+
+// Given type of LHS and the operation finds the expected type of RHS.
+// The type will be the same as LHS for simple operations like =, but it will be different for more complex ones like IN or CONTAINS.
+static lw_shared_ptr<column_specification> get_rhs_receiver(lw_shared_ptr<column_specification>& lhs_receiver, oper_t oper) {
+    const data_type& lhs_type = lhs_receiver->type->underlying_type();
+
+    if (oper == oper_t::IN) {
+        data_type rhs_receiver_type = list_type_impl::get_instance(std::move(lhs_type), false);
+        auto in_name = ::make_shared<column_identifier>(format("in({})", lhs_receiver->name->text()), true);
+        return make_lw_shared<column_specification>(lhs_receiver->ks_name,
+                                                    lhs_receiver->cf_name,
+                                                    in_name,
+                                                    std::move(rhs_receiver_type));
+    } else if (oper == oper_t::CONTAINS) {
+        if (lhs_type->is_list()) {
+            return list_value_spec_of(*lhs_receiver);
+        } else if (lhs_type->is_set()) {
+            return set_value_spec_of(*lhs_receiver);
+        } else if (lhs_type->is_map()) {
+            return map_value_spec_of(*lhs_receiver);
+        } else {
+            throw exceptions::invalid_request_exception(format("Cannot use CONTAINS on non-collection column \"{}\"",
+                                                                     lhs_receiver->name));
+        }
+    } else if (oper == oper_t::CONTAINS_KEY) {
+        if (lhs_type->is_map()) {
+            return map_key_spec_of(*lhs_receiver);
+        } else {
+            throw exceptions::invalid_request_exception(format("Cannot use CONTAINS KEY on non-map column {}",
+                                                               lhs_receiver->name));
+        }
+    } else if (oper == oper_t::LIKE) {
+        if (!lhs_type->is_string()) {
+            throw exceptions::invalid_request_exception(
+                format("LIKE is allowed only on string types, which {} is not", lhs_receiver->name->text()));
+        }
+        return lhs_receiver;
+    } else {
+        return lhs_receiver;
+    }
+}
+
+static expression prepare_binop_rhs(const expression& rhs, const lw_shared_ptr<column_specification>& rhs_receiver, oper_t oper, data_dictionary::database db, const sstring& keyspace) {
+    // Lists of IN values must be handled separately during preparation
+    // because they might contain nulls and bind markers.
+    if (oper == oper_t::IN && is<collection_constructor>(rhs)) {
+        const collection_constructor& rhs_in_list = as<collection_constructor>(rhs);
+        return list_prepare_expression(rhs_in_list, db, keyspace, rhs_receiver, true);
+    }
+
+    return prepare_expression(rhs, db, keyspace, rhs_receiver);
```

Comment:
I think this highlights the bug in lists::operations::setter. In general prepare_expression checks for NULLs in collection_constructor, but there is no such check in collections received via bind variables. Or maybe there is and it is well hidden.

Reply:
You're right, there doesn't seem to be a proper check for null in a collection received from bind marker.
`evaluate(const bind_variable&)` Calls `validate` on the type, but `validate` doesn't check for nulls.
A check could be added in `validate()`.
In case of lists there could theoretically be nulls in case of lists of IN values, but this doesn't work at the moment so maybe we don't need to support it at all.

Actually now that I checked it cassandra doesn't allow for NULLs in an IN list anymore :O
We could forbid nulls there and we are still gonna be compatible with cassandra.
This will solve the `evaluate_IN_list` problem altogether :)
I will open an issue for it.

Trying to send a collection with null values inside fails but in an ugly way.
`validate()` does `read_collection_value` for all elements, which reads element size, interprets it as unsigned int and reads that much. In case of `null` value this size is `-1` so it tries to read some random huge number of bytes and then fails.

Reply:
Opened #10579 and #10580